packages/astro/src/core/build/plugins/plugin-ssr.ts

```diff
@@ -71,12 +73,216 @@ function vitePluginSSR(
+				return '\0' + id;
+			}
+		},
+		async load(id) {
+			if (id.startsWith(RESOLVED_SERVERLESS_MODULE_ID)) {
+				const {
+					settings: { config },
+					allPages,
+				} = options;
+				const imports: string[] = [];
+				const contents: string[] = [];
+				const exports: string[] = [];
+
+				const path = getPathFromVirtualModulePageName(RESOLVED_SERVERLESS_MODULE_ID, id);
+				const virtualModuleName = getVirtualModulePageNameFromPath(ASTRO_PAGE_MODULE_ID, path);
+				let module = await this.resolve(virtualModuleName);
+				if (module) {
+					// we need to use the non-resolved ID in order to resolve correctly the virtual module
+					imports.push(`import * as pageModule from "${virtualModuleName}";`);
+				}
+
+				const ssrCode = generateSSRCode(options.settings.config, adapter);
+				imports.push(...ssrCode.imports);
+				contents.push(...ssrCode.contents);
+
+				return `${imports.join('\n')}${contents.join('\n')}${exports.join('\n')}`;
+			}
+			return void 0;
+		},
+		async generateBundle(_opts, bundle) {
+			// Add assets from this SSR chunk as well.
+			for (const [_chunkName, chunk] of Object.entries(bundle)) {
+				if (chunk.type === 'asset') {
+					internals.staticFiles.add(chunk.fileName);
+				}
+			}
+
+			for (const [chunkName, chunk] of Object.entries(bundle)) {
+				if (chunk.type === 'asset') {
+					continue;
+				}
+				let shouldDeleteBundle = false;
+				for (const moduleKey of Object.keys(chunk.modules)) {
+					if (moduleKey.startsWith(RESOLVED_SERVERLESS_MODULE_ID)) {
+						internals.ssrServerlessEntryChunks.set(moduleKey, chunk);
+						shouldDeleteBundle = true;
+					}
+				}
+				if (shouldDeleteBundle) {
+					delete bundle[chunkName];
+				}
+				// if (chunk.modules[RESOLVED_SSR_VIRTUAL_MODULE_ID]) {
+				// }
+			}
+		},
+	};
+}
+
+export function pluginSSRServerless(
+	options: StaticBuildOptions,
+	internals: BuildInternals
+): AstroBuildPlugin {
+	const ssr = isServerLikeOutput(options.settings.config);
+	return {
+		build: 'ssr',
+		hooks: {
+			'build:before': () => {
+				let vitePlugin =
+					ssr && options.settings.config.build.ssrMode === 'serverless'
+						? vitePluginSSRServerless(internals, options.settings.adapter!, options)
+						: undefined;
+
+				return {
+					enforce: 'after-user-plugins',
+					vitePlugin,
+				};
+			},
+			'build:post': async ({ mutate }) => {
+				if (!ssr) {
+					return;
+				}
+				if (options.settings.config.build.ssrMode === 'server') {
+					return;
+				}
+
+				if (internals.ssrServerlessEntryChunks.size === 0) {
+					throw new Error(`Did not generate an entry chunk for SSR serverless`);
+				}
+				// TODO: check this commented code
+				// Mutate the filename
+				// internals.ssrEntryChunk.fileName = options.settings.config.build.serverEntry;
+
+				for (const [moduleName, chunk] of internals.ssrServerlessEntryChunks) {
+					const code = await injectManifest(options, internals, chunk);
+					mutate(chunk, 'server', code);
+				}
+			},
```

Comment:
Does this mean we inject a manifest for each page chunk? It would be nice if they share a chunk that solely contains the manifest, if that's possible 🤔 

The difference between `SSRServerManifest` and `SSRServerlessManifest` seems to be the module instance value `pageModule` vs `pageMap`. If we can share a chunk that uses `pageMap` only, then each page chunk could manually load their own page module from that map?

Reply:
> I mean the map should contain all the pages, but the chunk would access only one page from the map. There shouldn't be a huge implication since each module is a dynamic import? (Or maybe I'm missing the idea of splitting here)

When classic SSR, the `entry.mjs` file will have something like this (pseudo-code, simplified):

```js
const page0 = () => import("chunks/page0.0303030.mjs");
const page1 = () => import("chunks/page1.04ofio3030.mjs");

const pageMap = new Map([
	["src/pages/index.astro", page0],
	["src/pages/blog.astro", page1],
])

const manifest = {
	pageMap
}
```

Instead, with SSR split, each file will look like this (pseudo-code, simplified):

```js
// file dist/pages/entry.index.astro.mjs
const pageModule = () => import("chunks/page0.0303030.mjs").then(n => n.i); 

const manifest = {
	pageModule
}
```

```js
// file dist/pages/entry.blog.astro.mjs
const pageModule = () => import("chunks/page1.04ofio3030.mjs").then(n => n.i); 

const manifest = {
	pageModule
}
```

That's why the `pageMap` and `pageModule`. What do you think? Should we keep `pageMap` for both?